# Библиотека PANDAS
### Борунов Семён Сергеевич НММбд-01-22

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В каждом разделе вам предлагается выполнить несколько заданий. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотекам pandas и matplotlib.

### Оценивание и штрафы

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написаем production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

### Формат сдачи
Для сдачи задания получившийся файл \*.ipynb с решением необходимо выложить в свой репозиторий github.

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных. С этой библиотекой мы сейчас и познакомимся.

Основной объект в pandas это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [1]:
%pylab inline  
# import almost all we need
import pandas as pd

%pylab is deprecated, use %matplotlib inline and import the required libraries.
Populating the interactive namespace from numpy and matplotlib


Данные находятся в приложенном файле `chipotle.tsv`

#### 1. Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

In [2]:
df = pd.read_csv("chipotle.tsv", sep="\t")
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

Это что-то про еду, заказы. Тут есть номер заказа, количество, название товара, описание и цена за все это удовольствие.

#### 2.  Ответьте на вопросы:
1. Сколько заказов попало в выборку?
2. Сколько уникальных категорий товара было куплено? (item_name)

In [3]:
len(df["item_name"].unique())

50

Заказов у нас 1834, это было видно еще в **1.**, товаров уникальных 50.

#### 3. Есть ли в данных пропуски? В каких колонках? 

In [4]:
df.dropna()

order_id  quantity           item_name  \
1            1         1                Izze   
2            1         1    Nantucket Nectar   
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
...        ...       ...                 ...   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[3376 rows x 5 columns]

Да, пропуски имеются, их 4621 - 3376 = 1245 штук. 

Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [5]:
df = df.fillna('')

#### 4. Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [6]:
price = df['item_price'].apply(lambda s: float(s[1:].strip()))
price

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

Какая средняя/минимальная/максимальная цена у товара? 

In [7]:
price.describe()

count    4622.000000
mean        7.464336
std         4.245557
min         1.090000
25%         3.390000
50%         8.750000
75%         9.250000
max        44.250000
Name: item_price, dtype: float64

**минимальное**: 1.09  
**максимальное**: 44.25  
**среднее**: 7.464336

удалите прошлую строку с ценами

In [8]:
df['item_price'] = price
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  item_price  
0                                                              2.39  
1                                          [Clementine]        3.39  
2                                               [Apple]        3.39  
3                                                              2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
...                                                 ...         ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...        8.75  

[4622 rows x 5 columns]

колонку не просто удалим, а заменим прошлую

#### 5. Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по item_name)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

In [9]:
df['item_one_price'] = df['item_price'] / df['quantity']
df.drop_duplicates('item_name').sort_values(by='item_one_price')[:5]

order_id  quantity          item_name choice_description  item_price  \
34         17         1      Bottled Water                           1.09   
18          9         2        Canned Soda           [Sprite]        2.18   
263       114         1  Canned Soft Drink             [Coke]        1.25   
6           3         1      Side of Chips                           1.69   
40         19         1              Chips                           2.15   

     item_one_price  
34             1.09  
18             1.09  
263            1.25  
6              1.69  
40             2.15

Это пятерка самых дешевых

In [10]:
df.drop_duplicates('item_name').sort_values(by='item_one_price')[:-6:-1]

order_id  quantity              item_name  \
281        123         2       Steak Salad Bowl   
1229       501         1    Barbacoa Salad Bowl   
1132       468         1    Carnitas Salad Bowl   
168         75         1  Barbacoa Crispy Tacos   
7            4         1          Steak Burrito   

                                     choice_description  item_price  \
281   [Tomatillo Red Chili Salsa, [Black Beans, Chee...       23.78   
1229  [Fresh Tomato Salsa, [Rice, Fajita Vegetables,...       11.89   
1132  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...       11.89   
168   [Tomatillo Red Chili Salsa, [Rice, Black Beans...       11.75   
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75   

      item_one_price  
281            11.89  
1229           11.89  
1132           11.89  
168            11.75  
7              11.75

и самые лакшери

#### 6. Сколько раз клиенты покупали больше 1 Chicken Bowl (item_name)?

In [11]:
df[(df['item_name'] == 'Chicken Bowl') & (df['quantity'] > 1)].head()

order_id  quantity     item_name  \
4           2         2  Chicken Bowl   
154        70         2  Chicken Bowl   
282       124         2  Chicken Bowl   
409       178         3  Chicken Bowl   
415       181         2  Chicken Bowl   

                                    choice_description  item_price  \
4    [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98   
154  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       17.50   
282  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...       17.50   
409  [[Fresh Tomato Salsa (Mild), Tomatillo-Green C...       32.94   
415                        [Tomatillo Red Chili Salsa]       17.50   

     item_one_price  
4              8.49  
154            8.75  
282            8.75  
409           10.98  
415            8.75

так сработало (с ```and``` не работало), осталось посчитать сколько в этой табличке строк

In [12]:
len(df[(df['item_name'] == 'Chicken Bowl') & (df['quantity'] > 1)])

33

#### 7. Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [13]:
df.groupby(by='order_id')['item_price'].sum().mean()

18.81142857142869

#### 8. Сколько заказов содержали ровно 1 товар?

In [14]:
sum(df.groupby(by='order_id')['quantity'].sum() == 1)

56

#### 9. Какая самая популярная категория товара? 

In [15]:
df.groupby('item_name')['quantity'].sum().sort_values()[:-4:-1]

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

Абсолютная победа у ```Chicken Bowl``` с отрывом в 170 покупок

#### 10. Какие виды Burrito существуют? Какой из них чаще всего покупают? Какой из них самый дорогой? 

In [16]:
only_burritos = df.mask(df['item_name'].apply(lambda s: 'Burrito' not in s)).dropna()
only_burritos['item_name'].unique()

array(['Steak Burrito', 'Chicken Burrito', 'Barbacoa Burrito',
       'Carnitas Burrito', 'Veggie Burrito', 'Burrito'], dtype=object)

Вот такие ~~пироги~~ буррито имеются  
Самое популяроное это ```Chicken Burrito```, это мы увидели еще в **9.**, но в целом можно посмотреть это тем же способом, но для фрейма `only_burritos`.   
Так же и про самый дорогой видно в **5.**

#### 11. В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

Возможно, будет удобно посчитать отдельно среднюю стоимость заказа, добавить ее в исходные данные и сделать необходимые проверки.

*Данный комментарий стоит воспринимать как подсказку к одному из вариантов решений задания. Если в вашем варианте решения он не нужнен, это не страшно*

In [22]:
sum(df.groupby(by='order_id')['item_price'].transform(lambda p: (p/p.sum()) > 0.4))

2087

#### 12. Предположим, что в данных была ошибка и Diet Coke (choice_description), который стоил $1.25, должен был стоить 1.35. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [18]:
start_sum = df[df['choice_description'].str.contains('Diet Coke') & (df['item_one_price'] == 1.25)]['item_price'].sum()
start_sum

143.75

очевидно, что если нам нужно отношение прибылей, то оно не зависит от количества купленных товаров, ведь 
$$s_1 = p_1 q \\
s_2 = p_2 q \\
\frac{s_1}{s_2} = \frac{p_1 q}{p_2 q } = \frac{p_1}{p_2}$$  
ну и это отношение равно 1.08, те разница в доходах 8%

#### 13. Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [19]:
m = np.array([[1, 2, 3],
             [4, 5, 6],
             [7, 8, 9]])

mat = pd.DataFrame(m, columns=['index', 'column1', 'column2'])

# Установка первой колонки в качестве индекса
mat.set_index('index', inplace=True)

mat

column1  column2
index                  
1            2        3
4            5        6
7            8        9

Сохраните DataFrame на диск в формате csv без индексов и названий столбцов.

In [20]:
mat.to_csv('data.csv', index=False, header=False)

## 2. Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. В этом разделе мы этим и займёмся.

У matplotlib, конечно, же есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторое величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить каждый из них. Не менее важно усвоить базовые принципы визуализаций:
- на графиках должны быть подписаны оси;
- у визуализации должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах).
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать.

In [21]:
%matplotlib inline  # нужно для отображения графиков внутри ноутбука
import matplotlib.pyplot as plt

UsageError: unrecognized arguments: # нужно для отображения графиков внутри ноутбука


На самом деле мы уже импортировали matplotlib внутри %pylab inline в начале задания.

Работать мы будем с той же выборкой покупкок. Добавим новую колонку с датой покупки.

In [ ]:
import datetime

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pd.DataFrame(index=df.order_id.unique())
dates['date'] = [
    (start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
    for _ in range(df.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
df['date'] = df.order_id.map(dates['date'])
df

#### 14. Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов item_name. 

Изображайте на двух соседних графиках. Для этого может быть полезен subplot.

In [ ]:
fig, axs = plt.subplots(1, 2, figsize=(12, 6))


average_order_price = df['order_id'].map(df.groupby('order_id')['item_one_price'].sum())
# print(average_order_price)
axs[0].hist(average_order_price, bins=20, color='red')
axs[0].set_title('Распределение сумм покупок')
axs[0].set_xlabel('Сумма покупки')
axs[0].set_ylabel('Частота')

# Гистограмма средних цен отдельных видов продуктов
axs[1].hist(df.groupby('item_name')['item_one_price'].mean(), bins=20, color='green')
axs[1].set_title('Средние цены')
axs[1].set_xlabel('Средняя цена')
axs[1].set_ylabel('Частота')


plt.show()

#### 15. Постройте график зависимости суммы покупок от дней.

In [ ]:
df['date'] = pd.to_datetime(df['date'])
day = df.groupby(df['date'].dt.date)['item_price'].sum()

plt.figure(figsize=(12, 6))
plt.plot(day.index, day.values, linestyle='--')
plt.xlabel('День')
plt.ylabel('Сумма покупки')
plt.grid(True)
plt.show()

#### 16. Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
import calendar

plt.figure(figsize=(12, 6))

df['date'] = pd.to_datetime(df['date'])
df['day_of_week'] = df['date'].dt.day_name()

week_days = list(calendar.day_name)

price_per_day = df.groupby('day_of_week')['item_one_price'].mean()

price_per_day.reindex(week_days).plot(kind='bar', color='green')
plt.ylabel('Средняя цена покупки')
plt.show()

#### 17. Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
plt.figure(figsize=(12, 6))
plt.scatter(df['quantity'], df['item_one_price'], color='k')
plt.xlabel('Количество')
plt.ylabel('Цена товара')
plt.grid(True)
plt.show()

#### 18. Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
plt.figure(figsize=(12, 6))
plt.scatter(df['quantity'], df['item_one_price'], color='k')
plt.xlabel('Количество')
plt.ylabel('Цена товара')
plt.grid(True)


plt.savefig('график зависимости.pdf')

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё. 